### PR TITLE
[for team review] RETURNING for replicated tables

### PR DESCRIPTION
**Pull request by remotefish**

### CODE REVIEW

File: src/backend/executor/nodeModifyTable.c [hunk 1]
```diff
@@ -1386,11 +1387,18 @@ ExecModifyTable(ModifyTableState *node)
...
+			node->ps.plan->flow->locustype == CdbLocusType_Replicated;
+
 		/*
 		 * If we got a RETURNING result, return it to caller.  We'll continue
 		 * the work on next call.
 		 */
-		if (slot)
+		if (slot &&
+			(!explicitGatherMotion ||
+			 node->ps.plan->flow->segindex == GpIdentity.segindex))
```
Comment (remotefish) [inline src/backend/executor/nodeModifyTable.c]: better to do the filtering at motion level

File: src/backend/optimizer/plan/createplan.c [hunk 2]
```diff
@@ -5801,6 +5807,10 @@ adjust_modifytable_flow(PlannerInfo *root, ModifyTable *node)
 			else if (targetPolicyType == POLICYTYPE_REPLICATED)
 			{
 				/* Do nothing here, see main work in cdbpath_motion_for_join() */
+#if 1
+				all_subplans_entry = false;
+				replicated = true;
+#endif
```
Comment (remotefish) [inline src/backend/optimizer/plan/createplan.c]: mixed replicated and partitioned?

File: src/backend/executor/execMain.c [hunk 3]
```diff
@@ -4549,7 +4549,9 @@ FillSliceTable_walker(Node *node, void *context)
 		/* The gang beneath a Motion will be a reader. */
 		sendSlice->gangType = GANGTYPE_PRIMARY_READER;
 
-		if (sendFlow->flotype != FLOW_SINGLETON)
+		if (sendFlow->flotype != FLOW_SINGLETON &&
+			!(sendFlow->flotype == FLOW_REPLICATED &&
+			  sendFlow->locustype == CdbLocusType_Replicated))
```
Comment (remotefish) [inline src/backend/executor/execMain.c]: Is it necessary to also check `locustype`? Per my local tests check `flotype` is enough.

File: src/backend/executor/nodeMotion.c [hunk 4]
```diff
@@ -368,9 +369,17 @@ execMotionUnsortedReceiver(MotionState * node)
...
+		node->ps.plan->lefttree->flow &&
+		node->ps.plan->lefttree->flow->locustype == CdbLocusType_Replicated)
+		srcRoute = gp_singleton_segindex;
+	else
+		srcRoute = ANY_ROUTE;
+
 	recvRC = RecvTupleFrom(node->ps.state->motionlayer_context,
 						   node->ps.state->interconnect_context,
-						   motion->motionID, &tuple, ANY_ROUTE);
+						   motion->motionID, &tuple, srcRoute);
```
Comment (remotefish) [inline src/backend/executor/nodeMotion.c]: TODO: also copy the change in other callers to `RecvTupleFrom()`?

File: src/backend/optimizer/plan/createplan.c [hunk 5]
```diff
@@ -5748,6 +5754,7 @@ adjust_modifytable_flow(PlannerInfo *root, ModifyTable *node)
 			if (targetPolicyType == POLICYTYPE_PARTITIONED)
 			{
 				all_subplans_entry = false;
+				all_subplans_replicated = false;
```
Comment (pengzhout) [inline src/backend/optimizer/plan/createplan.c:5757]: Just a reminder, have we test the mixed subplans? replicated + entry or else?

File: src/backend/cdb/motion/cdbmotion.c [hunk 6]
```diff
@@ -667,8 +667,6 @@ RecvTupleFrom(MotionLayerState *mlStates,
 	}
 	else
 	{
-		Assert(pMNEntry->preserve_order != 0);
-
```
Comment (pengzhout) [inline src/backend/cdb/motion/cdbmotion.c]: what's this for, seems not related.

File: src/backend/optimizer/plan/createplan.c (hunk 5 above)
Comment (remotefish) [reply src/backend/optimizer/plan/createplan.c:5757]: Yes we ever verified this case and updated in the tracker: https://www.pivotaltracker.com/story/show/156714887/comments/189526724, if I understand the logic correctly multiple subplans can only be triggered with `INHERITS` tables or partitioned tables, which is not supported for replicated table, that means replicated + non-replicated is not possible yet.

File: src/backend/executor/execMain.c [hunk 7]
```diff
@@ -4549,7 +4549,8 @@ FillSliceTable_walker(Node *node, void *context)
 		/* The gang beneath a Motion will be a reader. */
 		sendSlice->gangType = GANGTYPE_PRIMARY_READER;
 
-		if (sendFlow->flotype != FLOW_SINGLETON)
+		if (sendFlow->flotype != FLOW_SINGLETON &&
+			sendFlow->flotype != FLOW_REPLICATED)
```
Comment (remotefish) [inline src/backend/executor/execMain.c]: Verified again, this change is not necessary. @pengzhout

File: src/backend/cdb/motion/cdbmotion.c (hunk 6 above)
Comment (remotefish) [reply src/backend/cdb/motion/cdbmotion.c]: When `RecvTupleFrom()` is called by `execMotionUnsortedReceiver()` it used to pass `ANY_ROUTE` and `pMNEntry->preserve_order = 0`, with the hack we'll pass a specific segment id for `srcRoute` then it can not pass the assert here.
